### DISCUSSION TITLE
At one moment Crate is starting to use a lot of RAM memory

### DISCUSSION THREAD

**Discussion by Ezaz09** [crate/crate#14952] [Q&A]
Hi. Me and my team is faced with some weird problem - at one moment Crate is starting to use a lot of RAM memory. At first time we thought that this problem is raised because at one moment our backed is launching a lot of scheduled jobs and in one time sending a lot of select sql queries and each of this queries provides a lot of records in results. But when we fixed this problems - made scheduled jobs launched not in one time and fixed select queries and also up limits for RAM memory which crate can use - crate just started to use much more RAM memory. And after some research we also noticed that this moment didn't depended on time when scheduled jobs is launching or when our application faces with large influx of customers. Socrates just starting to use a lot of RAM memory at random moment.

**Comment by matriv** [comment]
Hello @Ezaz09, 
- Do you have any logs of the CrateDB cluster at the moment this high mem usage occurs?
- What is the available RAM on your nodes? and what are the configured sizes for JVM?
- When you say lots of RAM usage, do you mean in the operating system, or in JVM, or both?
- Have you monitored CrateDB through [sys.jobs](https://cratedb.com/docs/crate/reference/en/5.4/admin/system-information.html#jobs) and [sys.jobs_metrics](https://cratedb.com/docs/crate/reference/en/5.4/admin/system-information.html#jobs-metrics)?
- When you say `also up limits for RAM memory` you mean you increased `CRATE_HEAP_SIZE` ?

**Reply by Ezaz09** [reply]
Hi @matriv, thanks for replying
1. Today we had the same problem with one of our node (we have 5 nodes up and running). There is logs, when this problem is started on the crate-1 node 
[crate-1 logs.txt](https://github.com/crate/crate/files/13162629/crate-1.logs.txt)
2. On the each node we spending 17 GB of RAM. When you asked about configured sizes of JVM, you mean sizes of JVM of applications which is sending sql queries to Crate or sizes of JVM of Crate?
3. I mean operating system. Our Crate is running in Kubernetes 
4. No we didn't check jobs in [sys.jobs](https://cratedb.com/docs/crate/reference/en/5.4/admin/system-information.html#jobs) and [sys.jobs_metrics](https://cratedb.com/docs/crate/reference/en/5.4/admin/system-information.html#jobs-metrics). We will do this when we faced with this problem in the next time
5. For `CRATE_HEAP_SIZE` we set 16 GB. No we didn't increase value of this variable

**Reply by matriv** [reply]
> On the each node we spending 17 GB of RAM. When you asked about configured sizes of JVM, you mean sizes of JVM of applications which is sending sql queries to Crate or sizes of JVM of Crate?

I meant for crate, so it's 16GB configured, thx.

1. What's the total mem on those machines? I'd expect something like 32GB as CrateDB apart from JVM memory needs also mem on the OS because of memory mapped files.
2. From the logs it seems that crate-1 tries to communicate with the other 2 nodes unsuccessfully, have you eliminated any networking issues? When this happens are those 2 nodes under high CPU load? Is there excessive GC running at that poing?

**Reply by Ezaz09** [reply]
1. Total mem is 48 GB, but all of this memory not only for Crate. How I mentioned before our Crate is up and running in Kubernetes
2. We are using cloud platform, similar to aws. So all our applications (backend, Crate and etc.) running on this platform, so that really weird that we are faced with some networking issues. I checked logs, this 2 nodes didn't was under high CPU load. About GC - we are using java 17, so we have default settings for gc

**Reply by matriv** [reply]
Thank you for your responses.
If you have other apps running on the same machines then if they consume most of the memory (48GB - 16GB for JVM) there is not enough  memory for CrateDB to work with the filesystem (memory mapped files).
Please monitor your JVM, regarding CPU usage, memory usage, and frequent GC runs, The latter might mean the 16GB are not enough for your case, and JVM is forced to run frequent and long Garbage collections to free up memory, needed to continue with operations.

**Reply by Ezaz09** [reply]
Ok, thanks for your answer. We will monitor our problem and after a few day I will post some updates in this thread

**Reply by Ezaz09** [reply]
@matriv Hello 
Me and my team checked how much our Crate is using heap size and as you can see, Crate don't use all of heap size which was provided for him.
![1 heap size](https://github.com/crate/crate/assets/51817051/87d79f1e-927e-42fd-8862-9509104756bb)
But after last incident, we checked graphics of how pods was used RAM memory and we found weird situation - 3 nodes was worked, I mean they changed count of using RAM memory, but 2 nodes didn't changed count of using RAM memory, they just frozen  on one amount of RAM. And we also didn't see any logs, which somehow regards to this situation
![crate-1 freezed](https://github.com/crate/crate/assets/51817051/21bc9a00-1d3e-47f9-8502-5cf83d09a24a)
![crate-0 worked](https://github.com/crate/crate/assets/51817051/6c00c49b-cd3b-42cc-86ee-0e3eeba2a49a)
![crate-2 worked](https://github.com/crate/crate/assets/51817051/42153f8d-2687-49c2-a04e-c08c74e45274)
There is some examples of working(second and third screenshot) and frozen(first screenshot) pods 
At the end of graphic of first screenshot you can see some jump, thats because of we rebooted this pod
And because of that our application stop working and we forced to reboot frozen pods. Reboot is fixing this problem
